### PR TITLE
Refactor RawConfig to make possible load envs from Json object

### PR DESCRIPTION
**Pull request by prushin**

### CODE REVIEW

File: src/LndClient/Data/LndEnv.hs [hunk 1]
```diff
@@ -1,58 +1,133 @@
...
+      Right r -> [(r, "")]
+      Left _ -> case reads x of
+        [(r, "")] -> [(r, "")]
+        _ -> []
+
+instance FromJSON LndPort where
+  parseJSON x =
+    case x of
+      A.Number s -> do
+        let maybeInt :: Maybe Int = toBoundedInteger s
```
Comment (tim2CF) [inline src/LndClient/Data/LndEnv.hs:101]: seems like `toBoundedInteger` function is polymorphic, so you can obtain there `Maybe Word32` so, don't extra `safeFromIntegral` computation in next line

File: src/LndClient/Data/LndEnv.hs [hunk 2]
```diff
@@ -1,58 +1,133 @@
...
+import LndClient.Import.External as Ex
 import LndClient.Util as U
 import Network.GRPC.HighLevel.Generated
 import Network.GRPC.LowLevel.Client
+import Text.Read
 
-newtype LndWalletPassword = LndWalletPassword ByteString
-  deriving (PersistField, PersistFieldSql, Eq)
+newtype LndWalletPassword = LndWalletPassword Text
+  deriving (PersistField, PersistFieldSql, Eq, FromJSON, IsString, Show)
```
Comment (tim2CF) [inline src/LndClient/Data/LndEnv.hs:36]: Oh man, please be careful with `Show` type class. We don't want any type of credentials or sensitive personal data to be able to be shown! It's really bad idea to make `LndWalletPassword ` and similar types showable. As future refactoring step we need to remove `Show` class from pretty much everywhere in our codebase.

I understand that you needed it for Hspec test, but let's do it properly. It's not easy question how to do it. There are so many ways to do it, this question deserves separate discussion.. We can implement `Show` instead of deriving it, showing for example harmless hash of data instead of data itself, or just show something like `(LndWalletPassword "SECRET"). We can create our own type class `ShowDebug` which is different from `Show` and use it in tests only and don't export from library in any form. We can create separate newtypes for test purposes. So many possibilities.
 
As shortest solution for now, we can refactor Hspec test a bit, something like 

```haskell
myRawEnv == fromJSON myRawEnvJSON `shouldBe` True
```

This will not give you nice output which parts are not matching if test fails, but there it's probably fine because so simple test will never fail. This refactoring will require to derive `Eq` which is really harmless if compare with `Show`

Comment (tim2CF) [reply src/LndClient/Data/LndEnv.hs:36]: Btw `Eq` is already derived there

File: src/LndClient/Data/LndEnv.hs (hunk 1 above)
Comment (prushin) [reply src/LndClient/Data/LndEnv.hs:101]: Okay)

File: src/LndClient/Data/LndEnv.hs (hunk 2 above)
Comment (prushin) [reply src/LndClient/Data/LndEnv.hs]: Simplified test, actually we don't need to check actual fields, but overall successfull parsing. So updated test, removed Show deriving and remove some constructors from LndEnv module export.

Comment (tim2CF) [reply src/LndClient/Data/LndEnv.hs]: For more complex cases actually I think this is the case where orphan instances might be useful. We define (derive) orphan instances of Show in our test folder, and export in tests where it's needed. Minimal effort, but we still have nice output in tests and don't expose Show instances from library

**Comment by tim2CF** [comment]
also remove and gitignore .DS_Store

File: src/LndClient/Data/LndEnv.hs [hunk 3]
```diff
@@ -62,69 +65,119 @@ data LndEnv
...
+        _ -> []
+
+instance FromJSON LndTlsCert where
+  parseJSON x =
+    case x of
+      A.String s ->
+        case createLndTlsCert $ C8.pack $ T.unpack s of
+          Right cert -> return cert
+          Left e -> failure e
+      e -> failure e
```
Comment (tim2CF) [inline src/LndClient/Data/LndEnv.hs:86]: It's nice to have some tests where `Read` and `FromJSON` instances are failing, just to check that our implementation is correct and smart constructor is applied. Not big deal, might be separate pull request.

File: src/LndClient/Data/LndEnv.hs [hunk 4]
```diff
@@ -62,69 +65,119 @@ data LndEnv
...
+      failure err = fail $ "Json port loading error: " <> " " <> show err
+
+instance FromJSON RawConfig where
+  parseJSON (Object v) =
+    RawConfig <$> v .: "lnd_wallet_password"
+      <*> v .: "lnd_tls_cert"
+      <*> v .: "lnd_hex_macaroon"
+      <*> v .: "lnd_host"
+      <*> v .: "lnd_port"
+  parseJSON _ = mzero
```
Comment (tim2CF) [inline src/LndClient/Data/LndEnv.hs:119]: Interesting, have never seen this kind of implementation, let's also write a test to see how `FromJSON` will fail if given JSON will be not an object, might be separate PR.
